Fixed #28520 -- Added --start-at/--start-after options to runtests.py.

hramezani opened the pull request:
[ticket 28520](https://code.djangoproject.com/ticket/28520)
Continue https://github.com/django/django/pull/8957 work.

felixxm commented on tests/runtests.py:
```diff
@@ -198,12 +198,22 @@ def no_available_apps(self):
+            elif start_after and (module_label == start_after or module_label.startswith(start_after + '.')):
+                found_start = True
+                continue
+            elif not found_start:
```
Checking `found_start` looks unnecessary, maybe: 
```python
else:
     continue
```

felixxm commented on tests/runtests.py:
```diff
@@ -198,12 +198,22 @@ def no_available_apps(self):
+        if not found_start:
+            if start_at and (module_label == start_at or module_label.startswith(start_at + '.')):
+                found_start = True
+            elif start_after and (module_label == start_after or module_label.startswith(start_after + '.')):
```
We use the same logic in at least three places, I would add internal hook, e.g. `_module_match_label(self, module_label, label)`.

felixxm commented on docs/internals/contributing/writing-code/unit-tests.txt:
```diff
@@ -213,6 +213,22 @@ Going beyond that, you can specify an individual test method like this:
    $ ./runtests.py --settings=path.to.settings i18n.tests.TranslationTests.test_lazy_objects
 
+Running tests from specific point
+---------------------------------
```
I don't think that a separate section is necessary here, keeping this under `Running only some of the tests` looks good for me.

felixxm commented on docs/internals/contributing/writing-code/unit-tests.txt:
```diff
@@ -213,6 +213,22 @@ Going beyond that, you can specify an individual test method like this:
+.. console::
+
+   $ ./runtests.py --start-after=wsgi
+
```
Can we add a note about how it works with `--reverse`? i.e. that it doesn't have any impact of these options. I was confusing for me, to be honest.

hramezani replied:
The `--reverse` option does not have any impact on these new options.
These options are for specifying the starting point of running tests. the `--reverse` option uses to reverse the order of the selected tests from the previous step.

felixxm commented on tests/runtests.py:
```diff
@@ -495,6 +513,17 @@ def paired_tests(paired_test, options, test_labels, parallel):
+                print('Aborting: --{} must be a top-level module.'.format(opt_display))
+                sys.exit(1)
+            else:
+                setattr(options, opt_name, os.path.normpath(opt_val))
```
@hramezani I think you removed `setattr(options, opt_name, os.path.normpath(opt_val))` by mistake. My proposition was to remove only `else`, i.e.
```python
if '.' in opt_val:
    print('Aborting: --%s must be a top-level module.' % opt_name.replace('_', '-'))
    sys.exit(1)
setattr(options, opt_name, os.path.normpath(opt_val))
```

hramezani replied:
Oh, Sorry, Fixed!

felixxm commented:
@hramezani I split commits and made minor fixes, but I also found a use case that doesn't work properly, e.g.:
```
./runtests.py --start-after=wsgi queries
```
`queries` is not in the list of modules after `wsgi` so it will not be added to the `INSTALLED_APPS` but in the same time we pass it to the `run_tests()` via `test_labels`:
```
failures = test_runner.run_tests(
    test_labels or get_installed(),
    extra_tests=extra_tests,
)
```

hramezani commented:
@felixxm Thanks for your help.
‍`queries` and any names that passed after it are considered as `test_labels` and ignores `get_installed()` in `test_labels or get_installed()`. 
We can check `test_labels` and remove items that don't exist in `get_installed()`. like this:
```
installed = get_installed()
    failures = test_runner.run_tests(
        [label for label in test_labels if label in installed] or installed,
        extra_tests=extra_tests,
    )
```

what is your idea?

felixxm commented:
@hramezani I'm not sure if we need `test_labels or get_installed()` at all :thinking:  because `setup()` is already changing the list of installed apps with `test_labels`.

hramezani commented:
Yes, you right I think. I think so.
do you mean to replace the `test_runner` like this?:

```
failures = test_runner.run_tests(
        get_installed(),
        extra_tests=extra_tests,
    )
```

felixxm commented:
@hramezani My mistake we need `test_labels` because it contains not only module names but also paths, tests names etc.

hramezani commented:
@felixxm, as you mentioned we need `test_labels`. so I think it is better to filter the first part of the `test_label`(app name) with the `installed app`.  like this:

```
installed = get_installed()
failures = test_runner.run_tests(
        [label for label in test_labels if label.split('.')[0] in installed] or installed,
        extra_tests=extra_tests,
    )
```
we have valid `installed_app` and `test_labels`

felixxm commented:
@hramezani I think specifying test labels and `--start-at/--start-after` should be mutually exclusive. Because even with proposed change all modules except `queries` [will be ignored](https://github.com/django/django/blob/480492fe70b0bb7df61c00854dc8535c9d21ba64/tests/runtests.py#L210-L214).

Moreover, currently implementation doesn't work with `--bisect` and `--pair`. Please fix this.

hramezani commented:
@felixxm Thanks. requested changes are done. I also added a [line](https://github.com/django/django/pull/11415/files#diff-5830044babd91b5beb690dd0104b4d56R231) to the doc about the using ``--start-at/--start-after`` and test labels together and rebased my branch.
Please let me know if anything else remains.

felixxm commented:
buildbot

felixxm commented:
buildbot

felixxm commented:
buildbot
